src/Components/Browser.JS/utils/package.json

```diff
@@ -0,0 +1,9 @@
+{
+  "name": "common",
+  "private": true,
+  "version": "0.0.1",
+  "description": "Common dependencies used during dev time. DO NOT PUBLISH",
+  "devDependencies": {
+    "rimraf": "^2.6.2"
+  }
+}
```

Comment:
I don't know why this is being added. Let's discuss.

Reply:
Another issue of the `file://` and maybe I'm just going about it wrong. On Windows, a reference with `file://` simply gets copied to the `node_modules` directory. Changing SignalR source files doesn't cause `yarn install` to invalidate or update it. The build currently cleans up the node_modules and runs `yarn install` as part building, this file exists to install `rimraf`

Reply:
> On Windows, a reference with file:// simply gets copied to the node_modules directory.

In that case it doesn't seem like a valid way of having a dependency at all. An NPM package, once published, isn't necessarily in the same shape as its pre-publish source files. We wouldn't know whether our packages are really going to work when published.

How pressing is our need to do source-to-source references for NPM packages?

 * If this is the only case, we could sidestep the issue entirely (at least for the forseeable future) by waiting for the SignalR package to be published (preview 6) then reverting to a package reference.
 * Longer term, if we can accept non-atomic updates, we could continue using package references but with whatever private NPM registry feed our CI pushes to. Some PRs would have to be two-phase, which I know isn't ideal, but keeps us completely in the standard happy path for NPM/Yarn.
 * If we absolutely have to have atomic repo-wide PRs, we could have the build publish NPM packages (`.tgz` files) to some artifacts location, presumably with a timestamp as the version, and reference them locally from there. It's not standard, but it's closer to standard than `file://`+hacks.

Reply:
> * If we absolutely have to have atomic repo-wide PRs, we could have the build publish NPM packages (`.tgz` files) to some artifacts location, presumably with a timestamp as the version, and reference them locally from there. It's not standard, but it's closer to standard than `file://`+hacks.

That is certainly an ideal way to do it, unfortunately my understanding is that it would take a lot of hacking Arcade to do it. I believe pack phase that generates the `.tgz` happens after the build phase for all the projects completes.

> How pressing is our need to do source-to-source references for NPM packages?

Since you embed the source directly, not doing a source-to-source reference would effectively mean you'd always be at least a build behind (since you'd only be able to use the publish outputs from the previous build at best). If we did the `.tgz` hack above we could get something a little cleaner and a little better, but I'm not sure what it would take to get there.

> An NPM package, once published, isn't necessarily in the same shape as its pre-publish source files.

This is true, but generally only in relation to `npmignore` files or `package.json`'s `files` element. NPM in fact requires that the source effectively be a superset of what is packaged, by the time `npm pack` is run. The only thing you can really do is filter **out** certain files via an exclude list (`npmignore`) or an include list ([`files`](https://docs.npmjs.com/files/package.json#files)).

Reply:
> Since you embed the source directly, not doing a source-to-source reference would effectively mean you'd always be at least a build behind

Yep, that's what I meant with "non-atomic updates" (sorry for being obscure).

While not perfect, I think this is a better situation to be in than fighting against both NPM/Yarn and Arcade and having weird hacks like extra `package.json` roots to work around restore issues.

Perhaps longer term we could request the functionality we need in Arcade to support multiple NPM packages properly (dropping `.tgz` artifacts as each one is built, so they can take package dependencies on each other). Or we go all-in on lerna or a similar alternative that addresses this in a sane way.

Reply:
> While not perfect, I think this is a better situation to be in

I'm not sure I agree. How would you get a coherent final build in this scenario? You'd be shipping 3.0 RTM while embedding the previous build of SignalR. In order to fix a Blazor-blocking bug in SignalR you'd either have to do two build cycles (somehow disabling Blazor in the first cycle) or manually flow the dependency (which feels hackier than what we're doing here). Not to mention the fact that we have extra lag introduced in even detecting possibly-breaking bugs. Blazor test failures could be due to a change introduced in an earlier build, causing confusion when investigating them.

It seems really odd to me to be shipping as a single product unit in the same repo and yet have dependent versions drifting between components within that same repo.

Also, it's not at all uncommon to have to use extra tools to manage this stuff. Angular's packages are [full of placeholder versions](https://github.com/angular/angular/blob/815443313044e0ec0ec87af588d09eda6b88f0c8/packages/platform-server/package.json#L15) that get replaced. They do things a little differently, but the premise is the same.

Reply:
OK, you're probably right that the drawbacks of that approach outweigh the benefits. I'll be interested to see how the alternative `link://`-based approach that @pranavkm has changed this to works out.